[GTK] Enable the async clipboard API

carlosgcampos opened the pull request:
#### ec3dd91419a8c035ddb2c79bbd85a93619bab1e9
<pre>
[GTK] Enable the async clipboard API
<a href="https://bugs.webkit.org/show_bug.cgi?id=211979">https://bugs.webkit.org/show_bug.cgi?id=211979</a>

Reviewed by Wenson Hsieh.

Enable async clipboard API and implement the required platform strategy
functions.

* LayoutTests/platform/gtk/TestExpectations:
* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebCore/Modules/async-clipboard/Clipboard.cpp:
(WebCore::Clipboard::ItemWriter::write):
(WebCore::Clipboard::ItemWriter::didSetAllData):
* Source/WebCore/Modules/async-clipboard/Clipboard.h:
* Source/WebCore/Modules/async-clipboard/ClipboardImageReader.cpp:
(WebCore::ClipboardImageReader::readBuffer):
* Source/WebCore/platform/Pasteboard.cpp:
(WebCore::Pasteboard::allPasteboardItemInfo const):
(WebCore::Pasteboard::pasteboardItemInfo const):
* Source/WebCore/platform/Pasteboard.h:
(WebCore::Pasteboard::name const):
* Source/WebCore/platform/PasteboardStrategy.h:
* Source/WebCore/platform/gtk/PasteboardGtk.cpp:
(WebCore::Pasteboard::Pasteboard):
(WebCore::Pasteboard::read):
(WebCore::Pasteboard::writeCustomData):
(WebCore::Pasteboard::changeCount const):
* Source/WebCore/platform/gtk/SelectionData.cpp:
(WebCore::SelectionData::clearAllExceptFilenames):
* Source/WebCore/platform/gtk/SelectionData.h:
(WebCore::SelectionData::addBuffer):
(WebCore::SelectionData::buffers const):
(WebCore::SelectionData::buffer):
(WebCore::SelectionData::clearBuffers):
* Source/WebKit/UIProcess/WebPasteboardProxy.cpp:
(WebKit::WebPasteboardProxy::writeCustomData):
* Source/WebKit/UIProcess/WebPasteboardProxy.h:
* Source/WebKit/UIProcess/WebPasteboardProxy.messages.in:
* Source/WebKit/UIProcess/gtk/Clipboard.h:
(WebKit::Clipboard::changeCount const):
* Source/WebKit/UIProcess/gtk/ClipboardGtk3.cpp:
(WebKit::Clipboard::Clipboard):
(WebKit::Clipboard::~Clipboard):
(WebKit::ReadURLAsyncData::ReadURLAsyncData):
(WebKit::Clipboard::readURL):
(WebKit::Clipboard::write):
* Source/WebKit/UIProcess/gtk/ClipboardGtk4.cpp:
(WebKit::Clipboard::write):
* Source/WebKit/UIProcess/gtk/WebPasteboardProxyGtk.cpp:
(WebKit::WebPasteboardProxy::writeToClipboard):
(WebKit::WebPasteboardProxy::writeCustomData):
(WebKit::pasteboardIemInfoFromFormats):
(WebKit::WebPasteboardProxy::allPasteboardItemInfo):
(WebKit::WebPasteboardProxy::informationForItemAtIndex):
(WebKit::WebPasteboardProxy::getPasteboardItemsCount):
(WebKit::WebPasteboardProxy::readURLFromPasteboard):
(WebKit::WebPasteboardProxy::readBufferFromPasteboard):
(WebKit::WebPasteboardProxy::getPasteboardChangeCount):
* Source/WebKit/WebProcess/WebCoreSupport/WebPlatformStrategies.cpp:
(WebKit::WebPlatformStrategies::changeCount):
* Source/WebKit/WebProcess/WebCoreSupport/WebPlatformStrategies.h:

Canonical link: <a href="https://commits.webkit.org/261740@main">https://commits.webkit.org/261740@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/4e58dafc2b0939a31e86e5d9b59bfad874c054cf

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [  ~~🧪 style~~](https://ews-build.webkit.org/#/builders/6/builds/112699 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/21856 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/90/builds/1371 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/87/builds/4475 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/121225 "Built successfully") 
| [  ~~🧪 bindings~~](https://ews-build.webkit.org/#/builders/11/builds/116768 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/23196 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/13021 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/86/builds/5634 "Passed tests") | 
| [  ~~🧪 webkitperl~~](https://ews-build.webkit.org/#/builders/19/builds/118468 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/17232 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/100468 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/105779 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/99183 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/88/builds/1012 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/46249 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/1/builds/100987 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/14176 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/89/builds/1052 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/95459 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/83/builds/12297 "Built successfully and passed tests") | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/14863 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/84/builds/10399 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/102488 "Built successfully") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/20193 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/53044 "Passed tests") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/46/builds/31994 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/8190 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/16708 "Built successfully") | | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/110530 "Built successfully") | 
| | | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/27296 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

aperezdc commented on Source/WebCore/Modules/async-clipboard/ClipboardImageReader.cpp:
```diff
@@ -27,15 +27,16 @@
+void ClipboardImageReader::readBuffer(const String&, const String&, Ref<SharedBuffer>&& buffer)
 {
-    notImplemented();
+    if (m_mimeType == "image/png"_s)
+        m_result = Blob::create(m_document.get(), buffer->extractData(), m_mimeType);
```
Do I understand correctly that this would also fix [bug #218519](https://bugs.webkit.org/show_bug.cgi?id=218519)?

carlosgcampos replied:
I don't think so

aperezdc reviewed (commented):
Without knowing the GTK clipboard API well, the changes look reasonable to me. Maybe @mcatanzaro or someone else who is comfortable with these parts of GTK can give the final approval.

whsieh commented on LayoutTests/platform/gtk/TestExpectations:
```diff
@@ -849,21 +849,18 @@ fast/forms/range/input-appearance-range-rtl.html [ ImageOnlyFailure ]
+editing/async-clipboard/sanitize-when-reading-markup.html [ Skip ]
+
+# These fail because are run with DOMPasteAllowed=false or JavaScriptCanAccessClipboard=false,
+# I don't know how they are supposed to work.
```
Nit - grammar ("These fail because are run").

These layout tests work by triggering programmatic paste (i.e. the page calls `document.execCommand("paste")` underneath user activation). On both macOS and iOS, this causes a contextual menu to be shown that gives the user an option to paste (thereby allowing the page to read from the system pasteboard). Since they're intended to exercise this programmatic paste mechanism, we explicitly turn `DOMPasteAllowed` off.

It sounds like we could make this work in GTK as a followup, but I think either skipping or marking these as failures/timeouts is the right call for now.

carlosgcampos replied:
Thanks for the clarification, I've changed the comment and updated the expectations to skip to avoid timeouts.

whsieh reviewed (approved):
The general clipboard-related (non-GTK-specific) changes LGTM!

(Might want to have another GTK reviewer take a look)

carlosgcampos committed:
[GTK] Enable the async clipboard API
https://bugs.webkit.org/show_bug.cgi?id=211979

Reviewed by Wenson Hsieh.

Enable async clipboard API and implement the required platform strategy
functions.

* LayoutTests/platform/gtk/TestExpectations:
* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebCore/Modules/async-clipboard/Clipboard.cpp:
(WebCore::Clipboard::ItemWriter::write):
(WebCore::Clipboard::ItemWriter::didSetAllData):
* Source/WebCore/Modules/async-clipboard/Clipboard.h:
* Source/WebCore/Modules/async-clipboard/ClipboardImageReader.cpp:
(WebCore::ClipboardImageReader::readBuffer):
* Source/WebCore/platform/Pasteboard.cpp:
(WebCore::Pasteboard::allPasteboardItemInfo const):
(WebCore::Pasteboard::pasteboardItemInfo const):
* Source/WebCore/platform/Pasteboard.h:
(WebCore::Pasteboard::name const):
* Source/WebCore/platform/PasteboardStrategy.h:
* Source/WebCore/platform/gtk/PasteboardGtk.cpp:
(WebCore::Pasteboard::Pasteboard):
(WebCore::Pasteboard::read):
(WebCore::Pasteboard::writeCustomData):
(WebCore::Pasteboard::changeCount const):
* Source/WebCore/platform/gtk/SelectionData.cpp:
(WebCore::SelectionData::clearAllExceptFilenames):
* Source/WebCore/platform/gtk/SelectionData.h:
(WebCore::SelectionData::addBuffer):
(WebCore::SelectionData::buffers const):
(WebCore::SelectionData::buffer):
(WebCore::SelectionData::clearBuffers):
* Source/WebKit/UIProcess/WebPasteboardProxy.cpp:
(WebKit::WebPasteboardProxy::writeCustomData):
* Source/WebKit/UIProcess/WebPasteboardProxy.h:
* Source/WebKit/UIProcess/WebPasteboardProxy.messages.in:
* Source/WebKit/UIProcess/gtk/Clipboard.h:
(WebKit::Clipboard::changeCount const):
* Source/WebKit/UIProcess/gtk/ClipboardGtk3.cpp:
(WebKit::Clipboard::Clipboard):
(WebKit::Clipboard::~Clipboard):
(WebKit::ReadURLAsyncData::ReadURLAsyncData):
(WebKit::Clipboard::readURL):
(WebKit::Clipboard::write):
* Source/WebKit/UIProcess/gtk/ClipboardGtk4.cpp:
(WebKit::Clipboard::write):
* Source/WebKit/UIProcess/gtk/WebPasteboardProxyGtk.cpp:
(WebKit::WebPasteboardProxy::writeToClipboard):
(WebKit::WebPasteboardProxy::writeCustomData):
(WebKit::pasteboardIemInfoFromFormats):
(WebKit::WebPasteboardProxy::allPasteboardItemInfo):
(WebKit::WebPasteboardProxy::informationForItemAtIndex):
(WebKit::WebPasteboardProxy::getPasteboardItemsCount):
(WebKit::WebPasteboardProxy::readURLFromPasteboard):
(WebKit::WebPasteboardProxy::readBufferFromPasteboard):
(WebKit::WebPasteboardProxy::getPasteboardChangeCount):
* Source/WebKit/WebProcess/WebCoreSupport/WebPlatformStrategies.cpp:
(WebKit::WebPlatformStrategies::changeCount):
* Source/WebKit/WebProcess/WebCoreSupport/WebPlatformStrategies.h:

Canonical link: https://commits.webkit.org/261740@main
